### PR TITLE
Fix multiple issues in `WorkerThreadPool`

### PR DESCRIPTION
**Pull request by RandomShaper**
- Fix project settings being ignored.
- Made usages of `native_thread_allocator` thread-safe.
- Remove redundant thread-safety from `low_priority_threads_used`, `exit_threads`.
- Fix deadlock due to unintended extra lock of `task_mutex`.

**NOTE:** I have a concern that by cherry-picking this into 4.0 the users that tweaked `threading/worker_pool/*` settings will now get a behavior change. It's the correct thing to happen, but maybe too surprising for 4.0, or at least warrants a note in the release notes.

**UPDATE:** For the records, this may also fix some hard to reproduce crash related to a thread object in one of the projects used to test #74405, that worked fine otherwise.

### CODE REVIEW

**Comment by akien-mga** [comment]
> **NOTE:** I have a concern that by cherry-picking this into 4.0 the users that tweaked `threading/worker_pool/*` settings will now get a behavior change. It's the correct thing to happen, but maybe too surprising for 4.0, or at least warrants a note in the release notes.

Doesn't seem too problematic. But yes, it will be mentioned in the changelog.

**Review by akien-mga — APPROVED** [review]

File: main/main.cpp [hunk 1]
```diff
@@ -1454,6 +1455,19 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 #endif
 	}
 
+	// Initialize WorkerThreadPool.
+	{
+		int worker_threads = GLOBAL_GET("threading/worker_pool/max_threads");
+		bool low_priority_use_system_threads = GLOBAL_GET("threading/worker_pool/use_system_threads_for_low_priority_tasks");
+		float low_property_ratio = GLOBAL_GET("threading/worker_pool/low_priority_thread_ratio");
+
+		if (editor || project_manager) {
```
Comment (RandomShaper) [inline main/main.cpp:1464]: I've force-pushed to amend this check. (Formerly, it was calling the getters in the `OS` singleton, but they still return `false` unconditionally at this point.)

**Comment by RandomShaper** [comment]
Sorry for the pushes, mark as draft, etc. I had just found an issue and thought it was a regression from this PR, but it isn't. I will just fix it in another PR, which may take more time.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Fantastic work!

**Comment by akien-mga** [comment]
This seems to be causing a deadlock when running unit tests `godot --test`, when compiling with `scons tests=yes` (or `dev_mode=yes`).

```
(gdb) bt
#0  0x00007f8821dfb736 in __futex_abstimed_wait_common () from /lib64/libc.so.6
#1  0x00007f8821dfdda8 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libc.so.6
#2  0x000000000a56c66c in Semaphore::wait (this=0xe8d16a8) at ./core/os/semaphore.h:56
#3  WorkerThreadPool::wait_for_task_completion (this=0xd6225c0, p_task_id=1) at core/object/worker_thread_pool.cpp:274
#4  0x0000000005987e62 in TestWorkerThreadPool::DOCTEST_ANON_FUNC_4977 () at ./tests/core/threads/test_worker_thread_pool.h:66
#5  0x00000000057b1b81 in doctest::Context::run (this=0x7ffd6311b9a0) at ./thirdparty/doctest/doctest.h:6930
#6  0x0000000005c2d145 in test_main (argc=2, argv=0x7ffd6311c018) at tests/test_main.cpp:186
#7  0x000000000576e168 in Main::test_entrypoint (argc=2, argv=0x7ffd6311c018, tests_need_run=@0x7ffd6311ba7f: true) at main/main.cpp:684
#8  0x00000000056f9e02 in main (argc=2, argv=0x7ffd6311c018) at platform/linuxbsd/godot_linuxbsd.cpp:55
```

**Comment by RandomShaper** [comment]
Quoting myself from RC:
> The problem is that, now the WorkerThreadPool is initialized once project settings are read, it's not initialized as part of the "very core" and, thus, it's not initialized for tests.
The simplest thing would be to add WorkerThreadPool::get_singleton()->init(); to test_main().
I'm not sure that is the right way to do it, though.
> In any case, I'm doing that in the PR. Should be fine at least for now.

**Comment by myaaaaaaaaa** [comment]
How late can `WorkerThreadPool` initialization be pushed to?

Specifically, if it can be initialized after `ScriptServer` (see below link), this can supersede #74501 and #76586 , which is currently blocking ThreadSanitizer in CI ( #73777 ) from being merged.

https://github.com/godotengine/godot/blob/fd4a06c51555904104b18494d0224f450d74fe2a/main/main.cpp#L2441

**Comment by myaaaaaaaaa** [comment]
Alternatively, `WorkerThreadPool::init()` and `WorkerThreadPool::finish()` can be merged into a single `WorkerThreadPool::resize()` function, defaulting to 0 worker threads [where tasks are run on the main thread](72704).

This would make `WorkerThreadPool` available for use immediately in single-threaded mode, allowing it to be safely "initialized" into multithreaded mode much later than was possible before.

It would also make it possible to unit test different `WorkerThreadPool` sizes.

**Comment by RandomShaper** [comment]
@myaaaaaaaaa, that can work. Could you please incorporate that change to your PR? Let's say that's a related but separate issue than the ones this PR aims to fix and it shouldn't take much longer to be merged. Aside, there should be a way to assert that the thread pool has been properly initialized with threads by some point. For instance, this PR was breaking tests by failing to initialize it because there would be zero threads to run the tasks, but at least we could realize. With the changes we are considering, the test cases would just happily run all the tasks on the main thread, which is not what they are meant to do.

**Comment by akien-mga** [comment]
Thanks!

**Comment by myaaaaaaaaa** [comment]
> @myaaaaaaaaa, that can work. Could you please incorporate that change to your PR? Let's say that's a related but separate issue than the ones this PR aims to fix and it shouldn't take much longer to be merged.

@RandomShaper  I would prefer to open a new PR for it after my outstanding threading PRs are merged/rejected first, to avoid headaches in development due to merge conflicts.

#73388

#72577
#72687
#72704
#72716
#72784

**Pull request merged**